### PR TITLE
Improve mesh editing force by line UI and fixes

### PR DESCRIPTION
**Pull request by vcloarec**
Following the PR #44962 that introduced the force by line tool for mesh editing, this PR aims to:
- fix force by line algorithm when layer CRS/ map CRS are different
- improve the UI and make it more consistent with other parts of QGIS.
- make the tool working with the new CAD Z value

To force by line, the user can either draw a line on the map canvas or select a vector layer feature on the map canvas by right click (same principle as select by polygon, but polyline can be selected).

There are several options that can be used to adapt the behavior. These settings are available in an action widget from the toolbar (existing in the first PR):
- whether vertices are added in the intersection of edge of faces with the forcing line
- either the new vertices have their Z value taken from the mesh (interpolated from the including face) or from the forcing line (interpolated between vertices of the line)
- Tolerance: if existing vertices are at a distance from the forcing line less than this tolerance, the existing vertices will be on result of the forcing line.

When new vertices have their Z value interpolated from the forcing line's vertices, the Z values of the forcing line depend on how the forcing line is entered.

When new vertices have their Z value interpolated from the forcing line's vertices, the Z values of the forcing line depend on how the forcing line is entered.

- forcing line drawn on the map canvas
each vertex will have the Z value of the Z value widget (on the top right of the map canvas) in standard mode or the Z value of the CAD dock widget in advanced digitizing mode.
If the mouse snaps on a mesh vertex or a vector layer feature vertex, Z value widget or CAD dock widget takes the value of snapped vertex and then will be used for the next vertex of the forcing line.

forceLine_1

- forcing line from an existing vector layer feature
If the geometry is 3D, the Z values of the geometry will be used. If not, the Z value will be the Z value of the Z value widget.

forceLine_2

### CODE REVIEW

File: src/app/mesh/qgsmaptooleditmeshframe.cpp [hunk 1]
```diff
@@ -356,22 +365,23 @@ void QgsMapToolEditMeshFrame::setActionsEnable( bool enable )
...
+      << mActionForceByLines
       << mActionReindexMesh;
 
   for ( QAction *action : std::as_const( actions ) )
     action->setEnabled( enable );
 
-  bool areGeometriesSelected = areGeometriesSelectedInVectorLayer();
-  mActionForceByVectorLayerGeometries->setEnabled( enable && areGeometriesSelected );
+//  bool areGeometriesSelected = areGeometriesSelectedInVectorLayer();
+//  mActionForceByLines->setEnabled( enable && areGeometriesSelected );
```
Comment (PeterPetrik) [inline src/app/mesh/qgsmaptooleditmeshframe.cpp:375]: remove?

**Review by PeterPetrik — APPROVED** [review]

File: src/app/mesh/qgsmaptooleditmeshframe.cpp [hunk 2]
```diff
@@ -2062,68 +2145,141 @@ bool QgsMapToolEditMeshFrame::isSelectionGrapped( QgsPointXY &grappedPoint )
...
+           vectorLayer->geometryType() == QgsWkbTypes::LineGeometry )
       {
-        QgsGeometry geom = feat.geometry();
+        QgsRectangle rect( x - sr, y - sr, x + sr, y + sr );
+        QgsCoordinateTransform transform = mCanvas->mapSettings().layerTransform( vectorLayer );
 
         try
         {
-          geom.transform( transform );
+          rect = transform.transformBoundingBox( rect, QgsCoordinateTransform::ReverseTransform );
```
Comment (elpaso) [inline src/app/mesh/qgsmaptooleditmeshframe.cpp:2219]: @PeterPetrik FYI this broke the builds on master.

Comment (PeterPetrik) [reply src/app/mesh/qgsmaptooleditmeshframe.cpp:2219]: I have seen you fixed it? Thanks for that... not sure since there was all green from CI

Comment (elpaso) [reply src/app/mesh/qgsmaptooleditmeshframe.cpp:2219]: No problem, it happened to me too to see all green and then have a failure... not sure why it happens.

Comment (PeterPetrik) [reply src/app/mesh/qgsmaptooleditmeshframe.cpp:2219]: not sure, maybe some change on the master in the meantime, but still without merge conflicts?

Comment (elpaso) [reply src/app/mesh/qgsmaptooleditmeshframe.cpp:2219]: that's possible, if ReverseTransform was moved to qgis.h and renamed.

**Pull request merged**